### PR TITLE
Add the filter pushdown on the conditions of timestamp and partitions to Kafka connector

### PR DESCRIPTION
**Pull request by yupeng9**
Part of the change is from this PR: https://github.com/trinodb/trino/commit/a2ec0e357c0c7301e38fa595233248ff1d9dc391

The change includes
 - Add the filter pushdown on the conditions of timestamp and partitions to Kafka connector
 - Add integration tests for kafka pushdown
 - Also reorganized the KafkaQueryRunner for set up

Test plan - (Please fill in how you tested your changes)
Unit tests and integration tests

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==
Kafka Changes
* Add the filter pushdown on the conditions of timestamp and partitions to Kafka connector
```

### CODE REVIEW

**Comment by rohanpednekar** [comment]
Hi @yangy0000 , do you think you can do a first pass review on this?

File: presto-docs/src/main/sphinx/connector/kafka.rst [hunk 1]
```diff
@@ -127,6 +129,20 @@ files (must end with ``.json``) which contain table description files.
...
 This property is optional; the default is ``etc/kafka``.
 
+``kafka.timestamp-upper-bound-force-push-down-enabled``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The upper bound predicate on ``_timestamp`` column
+is pushed down only for topics using ``LogAppendTime`` mode.
+
+For topics using ``CreateTime`` mode, upper bound push down must be explicitly
+allowed via ``kafka.timestamp-upper-bound-force-push-down-enabled`` config property
```
Comment (zhenxiao) [inline presto-docs/src/main/sphinx/connector/kafka.rst:139]: allowed -> enabled

File: presto-docs/src/main/sphinx/connector/kafka.rst [hunk 2]
```diff
@@ -48,18 +48,20 @@ Configuration Properties
...
+``kafka.table-names``                                      List of all tables provided by the catalog
+``kafka.default-schema``                                   Default schema name for tables
+``kafka.nodes``                                            List of nodes in the Kafka cluster
+``kafka.connect-timeout``                                  Timeout for connecting to the Kafka cluster
+``kafka.max-poll-records``                                 Maximum number of records per poll
+``kafka.max-partition-fetch-bytes``                        Maximum number of bytes from one partition per poll
+``kafka.table-description-dir``                            Directory containing topic description files
+``kafka.hide-internal-columns``                            Controls whether internal columns are part of the table schema or not
+``kafka.messages-per-split``                               Number of messages that are processed by each Presto split, defaults to 100000
+``kafka.timestamp-upper-bound-force-push-down-enabled``    Controls if upper bound timestamp push down is enabled for topics using ``CreateTime`` mode
```
Comment (zhenxiao) [inline presto-docs/src/main/sphinx/connector/kafka.rst:63]: shall we use `timestamp-upper-bound-push-down-enabled`? what do you think?

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConnectorConfig.java [hunk 3]
```diff
@@ -145,4 +156,29 @@ public KafkaConnectorConfig setHideInternalColumns(boolean hideInternalColumns)
         this.hideInternalColumns = hideInternalColumns;
         return this;
     }
+
+    public boolean isTimestampUpperBoundPushDownEnabled()
+    {
+        return timestampUpperBoundPushDownEnabled;
+    }
+
+    @Config("kafka.timestamp-upper-bound-force-push-down-enabled")
```
Comment (zhenxiao) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConnectorConfig.java:165]: timestamp-upper-bound-push-down-enabled?

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaFilterManager.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+            if (configMap != null) {
+                Config config = configMap.get(topicResource);
+                String timestampType = config.get(TOPIC_CONFIG_TIMESTAMP_KEY).value();
+                if (TOPIC_CONFIG_TIMESTAMP_VALUE_LOG_APPEND_TIME.equals(timestampType)) {
+                    return true;
+                }
+            }
+        }
+        catch (Exception e) {
```
Comment (zhenxiao) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaFilterManager.java:149]: is Exception too broad? could IOException better?

File: presto-kafka/src/main/java/com/facebook/presto/kafka/Range.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.lang.Math.min;
+
+public class Range
```
Comment (zhenxiao) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/Range.java:25]: is it possible to re-use:
com.facebook.presto.common.predicate.Range?

**Review by zhenxiao — COMMENTED** [review]
nice work, @yupeng9 
just left a few questions
could we squash the 3 commits into 1, with meaningful commit message?
If you are cherry-picking, could you please put cherry-pick message in each of the commit?

**Pull request closed without merging**